scene/gui/graph_edit.cpp

```diff
@@ -409,8 +409,14 @@ void GraphEdit::remove_child_notify(Node *p_child) {
 	if (gn) {
 		gn->disconnect("position_offset_changed", callable_mp(this, &GraphEdit::_graph_node_moved));
 		gn->disconnect("raise_request", callable_mp(this, &GraphEdit::_graph_node_raised));
-		gn->disconnect("item_rect_changed", callable_mp((CanvasItem *)connections_layer, &CanvasItem::update));
-		gn->disconnect("item_rect_changed", callable_mp((CanvasItem *)minimap, &GraphEditMinimap::update));
+
+		// In case of the whole GraphEdit being destroyed these references can already be freed.
+		if (connections_layer->is_inside_tree()) {
```

Comment:
This isn't right solution because address sanitizer still show usage of freed memory.
If `connections_layer` is already removed by e.g. `memdelete` we can't use it anymore, because at this place there may already be completely different data here.

Reply:
Yes, I've concidered something like this. In my tests it was still a reference to an object, just not a valid object. I initially wanted something more robust, so in `3.2` version I've tried using `ObjectDB::instance_validate`, but the result was mixed. The errors were gone, but the disconnects were still being performed, they just didn't err anymore. Almost like calling `ObjectDB::instance_validate` refreshed the references.

Plus, there is no equivalent to this in `master`. I've tried several other ways to evaluate if those variables still point somewhere reasonable, but had no luck. So I considered that I may be overthinking this and tried this, and it worked just as expected. But a better way may have escaped me due to my lack of experience with C++ and memory management, and the whole object system in Godot.

So what would be a right solution? I can for example cast to to the expected type and check that before checking for the inside the tree?
```cpp
CanvasItem *cl = Object::cast_to<CanvasItem>(connections_layer);
if (cl && cl->is_inside_tree()) {
  ...
```
And the same for the minimap. Would that be sufficient?

Reply:
Deleted variables should never be used except when given a value or nullptr.
That variable doesn't contains information in runtime if is valid or not(unlike e.g. in Java) so it is .

I tested that and this code doesn't show any errors and warnings but I'm not sure if logic is good:
```
	if (gn) {
		gn->disconnect("offset_changed", this, "_graph_node_moved");
		gn->disconnect("raise_request", this, "_graph_node_raised");
		if (is_inside_tree()) {
			gn->disconnect("item_rect_changed", connections_layer, "update");
			gn->disconnect("item_rect_changed", minimap, "update");
		}
	}
```

Reply:
This seems to be the same approach as was patched a couple of lines earlier for 
```cpp
top_layer->call_deferred("raise");
```

So I guess it works. 

Something more robust would probably be using the fact that `remove_child_notify` for those nodes would be called before it is called for `GraphNode`. So we can probably check if we are removing `connections_layer` and `top_layer` (for `minimap`) and give those variables `nullptr`. Then it would be a straightforward check here. What do you think about that?

Reply:
So, something like this:

```cpp

void GraphEdit::remove_child_notify(Node *p_child) {
	Control::remove_child_notify(p_child);

	if (p_child == top_layer) {
		top_layer = nullptr;
		minimap = nullptr;
	}
	else if (p_child == connections_layer) {
		connections_layer = nullptr;
	}

	if (top_layer != nullptr && is_inside_tree()) {
		top_layer->call_deferred("raise"); // Top layer always on top!
	}

	GraphNode *gn = Object::cast_to<GraphNode>(p_child);
	if (gn) {
		gn->disconnect("position_offset_changed", callable_mp(this, &GraphEdit::_graph_node_moved));
		gn->disconnect("raise_request", callable_mp(this, &GraphEdit::_graph_node_raised));

		// In case of the whole GraphEdit being destroyed these references can already be freed.
		if (connections_layer != nullptr && connections_layer->is_inside_tree()) {
			gn->disconnect("item_rect_changed", callable_mp((CanvasItem *)connections_layer, &CanvasItem::update));
		}
		if (minimap != nullptr && minimap->is_inside_tree()) {
			gn->disconnect("item_rect_changed", callable_mp((CanvasItem *)minimap, &GraphEditMinimap::update));
		}
	}
}
```

Don't see any issues with the editor or a running project with that so far.

Reply:
I tested code from previous comment, and now doesn't use invalid memory.